test/extended/networking/ipsec.go

```diff
@@ -0,0 +1,709 @@
+			err = e2epod.DeletePodWithWait(context.Background(), f.ClientSet, config.srcNodeConfig.tcpdumpPod)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			config.srcNodeConfig.tcpdumpPod = nil
+
+			err = e2epod.DeletePodWithWait(context.Background(), f.ClientSet, config.dstNodeConfig.pingPod)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			config.dstNodeConfig.pingPod = nil
+			err = e2epod.DeletePodWithWait(context.Background(), f.ClientSet, config.dstNodeConfig.tcpdumpPod)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			config.dstNodeConfig.tcpdumpPod = nil
+		}
+
+		checkForGeneveOnlyPodTraffic := func(config *testConfig) {
+			err := setupTestPods(config, false)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			defer func() {
+				// Don't cleanup test pods in error scenario.
+				if err != nil && !framework.TestContext.DeleteNamespaceOnFailure {
+					return
+				}
+				cleanupTestPods(config)
+			}()
+			err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, geneve)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, esp)
+			o.Expect(err).To(o.HaveOccurred())
+			err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, icmp)
+			o.Expect(err).To(o.HaveOccurred())
+		}
+
+		checkForESPOnlyPodTraffic := func(config *testConfig) {
+			err := setupTestPods(config, false)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			defer func() {
+				// Don't cleanup test pods in error scenario.
+				if err != nil && !framework.TestContext.DeleteNamespaceOnFailure {
+					return
+				}
+				cleanupTestPods(config)
+			}()
+			err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, esp)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, geneve)
+			o.Expect(err).To(o.HaveOccurred())
+		}
+
+		checkPodTraffic := func(mode v1.IPsecMode) {
+			if mode == v1.IPsecModeFull {
+				checkForESPOnlyPodTraffic(config)
+			} else {
+				checkForGeneveOnlyPodTraffic(config)
+			}
+		}
+
+		checkNodeTraffic := func(ipsecForNode bool) {
+			err := setupTestPods(config, true)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			defer func() {
+				// Don't cleanup test pods in error scenario.
+				if err != nil && !framework.TestContext.DeleteNamespaceOnFailure {
+					return
+				}
+				cleanupTestPods(config)
+			}()
+			if ipsecForNode {
+				err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, esp)
+				o.Expect(err).NotTo(o.HaveOccurred())
+				err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, icmp)
+				o.Expect(err).To(o.HaveOccurred())
+				return
+			} else {
+				err = pingAndCheckNodeTraffic(config.srcNodeConfig, config.dstNodeConfig, icmp)
+				o.Expect(err).NotTo(o.HaveOccurred())
+			}
+		}
+
+		checkTraffic := func(mode v1.IPsecMode, ipsecForNode bool) {
+			checkPodTraffic(mode)
+			checkNodeTraffic(ipsecForNode)
+		}
```

Comment:
I would get rid of this function because it just makes the test harder to understand.

Just have these two and use them appropriately on the tests
```
checkPodTraffic := func(mode v1.IPsecMode) ...
...
checkNodeTraffic := func(mode v1.IPsecMode) ...
...
```

Then on the E/W tests you would just call
```
checkPodTraffic(config.ipsecMode)
// this test does not enable N/S ipsec config, so node traffic behaves as it were disabled
checkNodeTraffic(v1.IPsecModeDisabled)
...
checkPodTraffic(mode)
// this test does not enable N/S ipsec config, so node traffic behaves as it were disabled
checkNodeTraffic(v1.IPsecModeDisabled)
```

On the N/S tests
```
checkPodTraffic(config.ipsecMode)
// N/S ipsec config is not in effect yet, so node traffic behaves as it were disabled
checkNodeTraffic(v1.IPsecModeDisabled)
...
checkPodTraffic(mode)
// N/S ipsec config is not in effect yet, so node traffic behaves as it were disabled
checkNodeTraffic(v1.IPsecModeDisabled)
...
// pod traffic will be encrypted as a result N/S encryption being enabled between this two nodes
checkPodTraffic(v1.IPsecModeFull)
checkNodeTraffic(mode)
```